src/wallet/tx_builder.cpp

```diff
@@ -1283,5 +1281,165 @@ cryptonote::transaction finalize_all_proofs_from_transfer_details(
     return tx;
 }
 //-------------------------------------------------------------------------------------------------------------------
+wallet2::pending_tx make_pending_carrot_tx(const carrot::CarrotTransactionProposalV1 &tx_proposal,
+    const wallet2::transfer_container &transfers,
+    const cryptonote::account_keys &acc_keys)
+{
+    const std::size_t n_inputs = tx_proposal.key_images_sorted.size();
+    const std::size_t n_outputs = tx_proposal.normal_payment_proposals.size() +
+        tx_proposal.selfsend_payment_proposals.size();
+    const bool shared_ephemeral_pubkey = n_outputs == 2;
+
```

Comment:
Early checks would be nice:

```suggestion
    CHECK_AND_ASSERT_THROW_MES(n_inputs > 0, "make_pending_carrot_tx: carrot tx proposal missing inputs");

```